### PR TITLE
chore: migrate to TypeScript (no runtime code changes)

### LINKED ISSUE #202: Add Support for TypeScript
Hi,

Are you willing to accept TypeScript typings within the repo itself?

This will make it more convenient for typescript users (and [suggested by typescript](https://github.com/DefinitelyTyped/DefinitelyTyped#create-a-new-package)) to provide typings with the repository itself, rather than another repo (such as [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped)).

Are you willing to have a pull request for this?

I will be happy to create typing for this library myself if you agree.

Thanks!

### LINKED ISSUE #277: Refactor everything to ES2016+ or better TypeScript
**Problem**
Much of the code is written in old style `es5`
But node 6+ already supports es2016 and up natively

**Describe the solution you'd like**
Convert all the code by hand into es2016

**Describe alternatives you've considered**
- Convert all code to es2016+ and use [babel](https://babeljs.io) to transpile the code down to es2016
- Convert all code to [TypeScript](http://www.typescriptlang.org) and compile the code down to es2016
TypeScript has the advantage of even better code optimization and documentation

**Additional context**
I have already started some successful experiments in this branch https://github.com/Shinigami92/HTMLHint/tree/experimental/babel-with-es2016
Also this branch is already a PR (https://github.com/thedaviddias/HTMLHint/pull/276): https://github.com/Shinigami92/HTMLHint/tree/refactoring/test/upgrade-to-es2016

[node support of ES201X](https://node.green)

### PR DESCRIPTION
**Pull request by Shinigami92**
***Short description of what this resolves:***

We want to create a stable code base in which errors can be identified quickly

***Proposed changes:***

- Rewrite the project with typescript

closes #202 
closes #277

### CODE REVIEW

**Comment by Shinigami92** [comment]
First experimental branch: https://github.com/htmlhint/HTMLHint/tree/chore/typescript-sub-1  

I created an initial process workflow to leave our current rollup build unchanged

**Comment by Shinigami92** [comment]
@thedaviddias I think I want to request a first review!

There are still `errors` and `warnings`, but from now on it might be that we need to **change** code!  
So we may push this into a beta branch?

_eslint_ is currently configured to allow warnings so the build can be triggered
The tests are green and I didn't touched the tests!

**Comment by Shinigami92** [comment]
For now I disabled some eslint rules and added some TODOs to resolve them later

This PR didn't touched runtime-code
This will be made in a next PR so we can track the changes in a better and more readable way

**Comment by Shinigami92** [comment]
I have improved linting for the project so test folder is also linted with typescript checking and I improved linting with type-checking enabled.

File: .eslintignore [hunk 1]
```diff
@@ -0,0 +1,12 @@
+node_modules
+
+# Automatically generated directories
+bin
+dist
+build
+
+# Website directory
+website
```
[the remark this answers is not in the document]
Comment (Shinigami92) [reply .eslintignore:9]: Currently I thought that the website is managed by docusaurus and therefore doesn't be linted
Isn't it something like a CMS?

File: .eslintrc.js [hunk 2]
```diff
@@ -22,5 +33,19 @@ module.exports = {
     quotes: ['error', 'single', { avoidEscape: true }],
     'template-curly-spacing': 'error',
     'template-tag-spacing': 'error',
+
+    // TODO: remove following disabled rules and use recommended
```
[the remark this answers is not in the document]
Comment (Shinigami92) [reply .eslintrc.js:37]: yes

File: build.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
+const { execSync } = require('child_process')
+
+const os = require('os')
+// control OS
+// then run command depengin on the OS
```
[the remark this answers is not in the document]
Comment (Shinigami92) [reply build.js:5]: I could remove them
This is copy-paste from StackOverflow!

File: .eslintignore (hunk 1 above)
[the remark this answers is not in the document]
Comment (thedaviddias) [reply .eslintignore:9]: After some thoughts I will probably create a eslint just inside the website repo

**Pull request merged**